collectors/redis_collector.go

```diff
@@ -1 +1,188 @@
+				Name:      "latency",
+				Help:      "Redis service latency in milliseconds",
+				ConstLabels: prometheus.Labels{
+					"environment": environment,
+				},
+			},
+			[]string{"service_guid", "plan_guid", "instance_guid"},
+		),
+	}
+}
+
+func (c *RedisCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.instanceAliveMetric.Describe(ch)
+	c.instanceHealthyMetric.Describe(ch)
+	c.instanceLatencyMetric.Describe(ch)
+}
+
+func (c *RedisCollector) Collect(ch chan<- prometheus.Metric) error {
+	c.instanceAliveMetric.Reset()
+	c.instanceHealthyMetric.Reset()
+	c.instanceLatencyMetric.Reset()
+
+	services, err := c.client.CoreV1().Services(c.namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("collector = %s", c.collectorLabel),
+	})
+	if err != nil {
+		return err
+	}
+
+	secrets, err := c.client.CoreV1().Secrets(c.namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("collector = %s", c.collectorLabel),
+	})
+	if err != nil {
+		return err
+	}
+
+	serviceSecrets := map[string][]apiv1.Secret{}
+	for _, secret := range secrets.Items {
+		serviceID := secret.ObjectMeta.Labels["service_id"]
+		serviceSecrets[serviceID] = append(serviceSecrets[serviceID], secret)
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, service := range services.Items {
+		wg.Add(1)
+		go func(service apiv1.Service) {
+			defer wg.Done()
+			values := map[string]string{}
+			for _, secret := range serviceSecrets[service.ObjectMeta.Labels["service_id"]] {
+				for key, value := range secret.Data {
+					values[key] = string(value)
+				}
+			}
+
+			ports := map[int32]int32{}
+			for _, port := range service.Spec.Ports {
+				ports[port.Port] = port.NodePort
+			}
+
+			alive, healthy, latency := c.checker(
+				fmt.Sprintf("%s.%s", service.Name, c.serviceDomain),
+				ports[6379],
+				values["redis-password"],
+			)
+
+			c.instanceAliveMetric.WithLabelValues(
+				service.ObjectMeta.Labels["catalog_service_id"],
+				service.ObjectMeta.Labels["catalog_plan_id"],
+				service.ObjectMeta.Labels["service_id"],
+			).Set(alive)
+			c.instanceHealthyMetric.WithLabelValues(
+				service.ObjectMeta.Labels["catalog_service_id"],
+				service.ObjectMeta.Labels["catalog_plan_id"],
+				service.ObjectMeta.Labels["service_id"],
+			).Set(healthy)
+			c.instanceLatencyMetric.WithLabelValues(
+				service.ObjectMeta.Labels["catalog_service_id"],
+				service.ObjectMeta.Labels["catalog_plan_id"],
+				service.ObjectMeta.Labels["service_id"],
+			).Set(latency)
+		}(service)
+	}
+	wg.Wait()
+
+	c.instanceAliveMetric.Collect(ch)
+	c.instanceHealthyMetric.Collect(ch)
+	c.instanceLatencyMetric.Collect(ch)
+
+	return nil
+}
+
+func CheckRedisHealth(host string, port int32, password string) (alive, healthy, latency float64) {
+	log.Infof("Sending PING to redis server at %s:%d", host, port)
+
+	client := redis.NewClient(&redis.Options{
+		Addr:     fmt.Sprintf("%s:%d", host, port),
+		Password: password,
+	})
+
+	start := time.Now()
+	pong, err := client.Ping().Result()
```

Comment:
WDYT about calling INFO here and basing the healthy check on if we hit a master / it has connected slaves?